zetta_utils/mazepa_layer_processing/common/subchunkable_apply_flow.py

```diff
@@ -0,0 +1,204 @@
+        return dst_resolution
+
+    def __call__(
+        self, idx: VolumetricIndex, dst: VolumetricLayer, *args: P.args, **kwargs: P.kwargs
+    ) -> None:
+        self.executor(self.flow_schema(idx, dst, *args, **kwargs))
+
+
+def expand_if_singleton(arg: Union[T, List[T]], length: int) -> List[T]:
+    if isinstance(arg, list):
+        if len(arg) == length:
+            return arg
+        else:
+            raise ValueError(f"cannot expand a list of length {len(arg)} to {length}")
+    return [arg for _ in range(length)]
+
+
+@builder.register(
+    "build_subchunkable_apply_flow",
+    cast_to_vec3d=["dst_resolution"],
+    cast_to_intvec3d=[
+        "fn_or_op_crop_pad",
+        "blend_pads",
+        "crop_pads",
+        "processing_chunk_sizes",
+        "max_reduction_chunk_sizes",
+    ],
+)
+def build_subchunkable_apply_flow(  # pylint: disable=keyword-arg-before-vararg, line-too-long, too-many-locals, too-many-branches
+    bcube: BoundingCube,
+    dst: VolumetricLayer,
+    dst_resolution: Vec3D,
+    temp_layers_dirs: Union[str, List[str]],
+    processing_chunk_sizes: List[IntVec3D],
+    crop_pads: Union[IntVec3D, List[IntVec3D]] = IntVec3D(0, 0, 0),
+    blend_pads: Union[IntVec3D, List[IntVec3D]] = IntVec3D(0, 0, 0),
+    blend_modes: Union[
+        Literal["linear", "quadratic"], List[Literal["linear", "quadratic"]]
+    ] = "linear",
+    fn_or_op_crop_pad: IntVec3D = IntVec3D(0, 0, 0),
+    fn: Optional[Callable[P, Tensor]] = None,
+    op: Optional[BlendableOpProtocol[P, Tensor]] = None,
+    max_reduction_chunk_sizes: Optional[Union[IntVec3D, List[IntVec3D]]] = None,
+    *args: P.args,
+    **kwargs: P.kwargs,
+) -> mazepa.Flow:
+
+    idx = VolumetricIndex(resolution=dst_resolution, bcube=bcube)
+
+    if fn is not None and op is not None:
+        raise ValueError("Cannot take both `fn` and `op`; please choose one or the other.")
+    if fn is None and op is None:
+        raise ValueError("Need exactly one of `fn` and `op`; received neither.")
+    if fn is not None:
+        level0_op = VolumetricCallableOperation(fn, crop_pad=fn_or_op_crop_pad)
+    elif op is not None:
+        # mypy seems to be confused by protocol
+        level0_op = op.with_added_crop_pad(fn_or_op_crop_pad)  # type:ignore
+    num_levels = len(processing_chunk_sizes)
+    if max_reduction_chunk_sizes is None:
+        max_reduction_chunk_sizes = processing_chunk_sizes
+
+    for arg in [
+        processing_chunk_sizes,
+        max_reduction_chunk_sizes,
+        crop_pads,
+        blend_pads,
+        blend_modes,
+        temp_layers_dirs,
+    ]:
+        if isinstance(arg, list):
+            if not len(arg) == num_levels:
+                raise ValueError(
+                    "The arguments `processing_chunk_sizes`, `max_reduction_chunk_sizes`"
+                    " (optional), `blend_pads`, `crop_pads`, `blend_modes`, `temp_layers_dirs`"
+                    " must be singletons or lists where the longest lengths match."
```

Comment:
Not sure what `lists where the longest lengths match` means here. Isn't this checking that all the lengths are equal to `processing_chunk_sizes`? Feels odd to include `processing_chunk_sizes` itself in the check.

Reply:
Oops, sorry - forgot to change that when I changed the code.